pkg/cloudprovider/providers/rackspace/rackspace.go

```diff
@@ -391,3 +467,145 @@ func (os *Rackspace) GetZone() (cloudprovider.Zone, error) {
+			glog.V(4).Infof("Disk: %q is already attached to compute: %q", diskName, instanceID)
+			return disk.ID, nil
+		}
+
+		errMsg := fmt.Sprintf("Disk %q is attached to a different compute: %q, should be detached before proceeding", diskName, disk.Attachments[0]["server_id"])
+		glog.Errorf(errMsg)
+		return "", errors.New(errMsg)
+	}
+
+	_, err = volumeattach.Create(compute, instanceID, &osvolumeattach.CreateOpts{
+		VolumeID: disk.ID,
+	}).Extract()
+	if err != nil {
+		glog.Errorf("Failed to attach %s volume to %s compute", diskName, instanceID)
+		return "", err
+	}
+	glog.V(2).Infof("Successfully attached %s volume to %s compute", diskName, instanceID)
+	return disk.ID, nil
+}
+
+func (rs *Rackspace) GetDevicePath(diskId string) string {
+	volume, err := rs.getVolume(diskId)
+	if err != nil {
+		return ""
+	}
+	attachments := volume.Attachments
+	if len(attachments) != 1 {
+		glog.Warningf("Unexpected number of volume attachements on %s: %d", diskId, len(attachments))
+		return ""
+	}
+	return attachments[0]["device"].(string)
+}
+
+// Takes a partial/full disk id or diskname
+func (rs *Rackspace) getVolume(diskName string) (volumes.Volume, error) {
+	sClient, err := rackspace.NewBlockStorageV1(rs.provider, gophercloud.EndpointOpts{
+		Region: rs.region,
+	})
+
+	var volume volumes.Volume
+	if err != nil || sClient == nil {
+		glog.Errorf("Unable to initialize cinder client for region: %s", rs.region)
+		return volume, err
+	}
+
+	err = volumes.List(sClient).EachPage(func(page pagination.Page) (bool, error) {
+		vols, err := volumes.ExtractVolumes(page)
+		if err != nil {
+			glog.Errorf("Failed to extract volumes: %v", err)
+			return false, err
+		}
+
+		for _, v := range vols {
+			glog.V(4).Infof("%s %s %v", v.ID, v.Name, v.Attachments)
+			if v.Name == diskName || strings.Contains(v.ID, diskName) {
+				volume = v
+				return true, nil
+			}
+		}
+
+		// if it reached here then no disk with the given name was found.
+		errmsg := fmt.Sprintf("Unable to find disk: %s in region %s", diskName, rs.region)
+		return false, errors.New(errmsg)
+	})
+	if err != nil {
+		glog.Errorf("Error occured getting volume: %s", diskName)
+	}
+	return volume, err
+}
+
+func (rs *Rackspace) getComputeClient() (*gophercloud.ServiceClient, error) {
+	client, err := rackspace.NewComputeV2(rs.provider, gophercloud.EndpointOpts{
+		Region: rs.region,
+	})
+	if err != nil || client == nil {
+		glog.Errorf("Unable to initialize nova client for region: %s", rs.region)
+	}
+	return client, nil
+}
+
+// Detaches given cinder volume from the compute running kubelet
+func (rs *Rackspace) DetachDisk(partialDiskId string) error {
+	instanceID, err := readInstanceID()
+	if err != nil {
+		return err
+	}
+
+	disk, err := rs.getVolume(partialDiskId)
+	if err != nil {
+		return err
+	}
+
+	compute, err := rs.getComputeClient()
+	if err != nil {
+		return err
+	}
+
+	if len(disk.Attachments) > 0 && instanceID == disk.Attachments[0]["server_id"] {
```

Comment:
I have objections about this: Rackspace's Public Cloud does not support "multiattach" (also see: https://developer.rackspace.com/docs/cloud-block-storage/v1/developer-guide/#document-general-api-info/cbsv1-methods-vs-cinderv2-methods)

```
$ export SERVER_A_ID=e6278a84-d1fe-4f65-b6e2-22cc391ae177
$ jq -n --arg volid $VOLUME_ID '{volumeAttachment:{device:"/dev/xvdb", volumeId: $volid}}' | http POST $COMPUTE_ENDPOINT/servers/$SERVER_A_ID/os-volume_attachments "X-Auth-Token: $TOKEN"
HTTP/1.1 200 OK
Content-Length: 195
Content-Type: application/json
Date: Thu, 14 Apr 2016 14:16:53 GMT
Date: Thu, 14 Apr 2016 14:16:54 GMT
Server: Jetty(9.2.z-SNAPSHOT)
Via: 1.1 Repose (Repose/7.1.3.0)
X-Compute-Request-Id: req-9a66ba75-9ffc-4299-9476-d3bdc735185a
x-trans-id: 852ab59d-f908-4e1e-b6f8-0492f7ba5dd9

{
    "volumeAttachment": {
        "device": "/dev/xvdb", 
        "id": "ae1c751d-efff-4bcc-bf16-a89493858c14", 
        "serverId": "e6278a84-d1fe-4f65-b6e2-22cc391ae177", 
        "volumeId": "ae1c751d-efff-4bcc-bf16-a89493858c14"
    }
}
```

```
$ export SERVER_B_ID=7faa4f6a-4029-44ce-b6e8-e9ef28e2b9c5
$ jq -n --arg volid $VOLUME_ID '{volumeAttachment:{device:"/dev/xvdb", volumeId: $volid}}' | http POST $COMPUTE_ENDPOINT/servers/$SERVER_B_ID/os-volume_attachments "X-Auth-Token: $TOKEN"
HTTP/1.1 400 Bad Request
Content-Length: 155
Content-Type: application/json; charset=UTF-8
Date: Thu, 14 Apr 2016 14:20:45 GMT
Date: Thu, 14 Apr 2016 14:20:45 GMT
Server: Jetty(9.2.z-SNAPSHOT)
Via: 1.1 Repose (Repose/7.1.3.0)
X-Compute-Request-Id: req-0b464867-9c13-4f5c-86d0-e294951edf92
x-trans-id: a6430fc0-dcef-4809-9e86-fdfe8aea41f0

{
    "badRequest": {
        "code": 400, 
        "message": "Invalid volume: volume 'ae1c751d-efff-4bcc-bf16-a89493858c14' status must be 'available'. Currently in 'in-use'"
    }
}
```

Furthermore I think there is an implicit assumption by k8s that volume mounts are exclusive, AFAIK the only supported file systems for those volumes are currently ext3/4. When there's shared block-storage over nodes, you'll run into problems. I don't think this should be in the domain of k8s, this is why there are glusterfs/ceph volume plugins.

If we want to be defensive/forward-compatible about this, we could bail out with an error when a volume has > 1 attachements (still, currently the Rackspace API does not allow it).

Makes sense?